core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -3099,9 +3100,23 @@ private void registerSetop(
 
     // A setop is in the same scope as its parent.
     scopes.put(call, parentScope);
-    for (SqlNode operand : call.getOperandList()) {
+    @NonNull SqlValidatorScope recursiveScope = parentScope;
+    if (enclosingNode.getKind() == SqlKind.WITH_ITEM) {
+      if (node.getKind() != SqlKind.UNION) {
```

Comment:
The recursive flag is not explicitly checked here because for only recursive queries we pass WITH_ITEM as enclosed node(essentially it is kind of an implicit check). For non recursive queries we pass the original enclosed node which is either WITH or its parent. 

`And I think we should implement this validation logic in org.apache.calcite.sql.validate.WithItemNamespace#validateImpl or org.apache.calcite.sql.validate.SetopNamespace#validateImpl instead of registerSetop. WDYT?`

As regards to the validation logic implementation I did go that path but I think code is getting messy. I thought if we can create a WithRecursive* classes for these we could do custom logic for recursive queries alone.